plugins/mmdblookup/mmdblookup.c

```diff
@@ -429,58 +475,105 @@ CODESTARTdoAction
+	// Extract and amend fields (to message) as configured.
+	for (int i = 0; i < pData->fieldList.nmemb; ++i) {
+		size_t field_len = strnlen((char *)(pData->fieldList.name[i]), 1024);
+		char field[field_len + 1];
+		strncpy(field, (char *)pData->fieldList.name[i], sizeof(field));
+		field[field_len] = '\0'; // Ensure null termination.
+
+		mmdb_err = mmdb_get_value(&result.entry, &entry_data, field, field_len);
+		if (MMDB_SUCCESS != mmdb_err) {
+			dbgprintf("Got an error looking up the entry data - %s\n", MMDB_strerror(mmdb_err));
+			ABORT_FINALIZE(RS_RET_OK);
+		}
+		if (!entry_data.has_data || entry_data.offset <= 0) {
+			continue;
+		}
 
-	if (MMDB_SUCCESS != status) {
-		dbgprintf("Got an error looking up the entry data - %s\n", MMDB_strerror(status));
-		ABORT_FINALIZE(RS_RET_OK);
-	}
+		value_len = 0;
+		// Decode the value.
+		switch (entry_data.type) {
+		case MMDB_DATA_TYPE_BOOLEAN:
+			value_len = snprintf(value, sizeof(value), "%d", entry_data.boolean);
+			break;
+		case MMDB_DATA_TYPE_UTF8_STRING:
+			value_len = snprintf(
+				value,
+				sizeof(value),
+				"%.*s",
+				entry_data.data_size,
+				entry_data.utf8_string
+			);
+			break;
+		case MMDB_DATA_TYPE_BYTES:
+			value_len = snprintf(
+				value,
+				sizeof(value),
+				"%.*s",
+				entry_data.data_size,
+				(const char *)entry_data.bytes
+			);
+			break;
+		case MMDB_DATA_TYPE_FLOAT:
+			value_len = snprintf(value, sizeof(value), "%.5f", entry_data.float_value);
+			break;
+		case MMDB_DATA_TYPE_DOUBLE:
+			value_len = snprintf(value, sizeof(value), "%.5f", entry_data.double_value);
+			break;
+		case MMDB_DATA_TYPE_UINT16:
+			value_len = snprintf(value, sizeof(value), "%" PRIu16, entry_data.uint16);
+			break;
+		case MMDB_DATA_TYPE_UINT32:
+			value_len = snprintf(value, sizeof(value), "%" PRIu32, entry_data.uint32);
+			break;
+		case MMDB_DATA_TYPE_INT32:
+			value_len = snprintf(value, sizeof(value), "%" PRIi32, entry_data.int32);
+			break;
+		case MMDB_DATA_TYPE_UINT64:
+			value_len = snprintf(value, sizeof(value), "%" PRIu64, entry_data.uint64);
+			break;
+		case MMDB_DATA_TYPE_UINT128: {
+			uint8_t *p = (uint8_t *)&entry_data.uint128;
+			value_len = snprintf(
+				value,
+				sizeof(value),
+				"0x"
+				"%02x%02x%02x%02x"
+				"%02x%02x%02x%02x"
+				"%02x%02x%02x%02x"
+				"%02x%02x%02x%02x",
+				p[0], p[1], p[2], p[3],
+				p[4], p[5], p[6], p[7],
+				p[8], p[9], p[10], p[11],
+				p[12], p[13], p[14], p[15]
+			);
+			break;
+		}
+		default:
+			dbgprintf("Got an error looking up the entry data - unknown data type\n");
+			ABORT_FINALIZE(RS_RET_OK);
+		}
 
-	size_t  memlen;
-	char   *membuf;
-	FILE   *memstream;
-	CHKmalloc(memstream = open_memstream(&membuf, &memlen));
+		if (value_len <= 0) {
+			continue;
+		}
 
-	if (entry_data_list != NULL && memstream != NULL) {
-		MMDB_dump_entry_data_list(memstream, entry_data_list, 2);
-		fflush(memstream);
-		str_split(&membuf);
-	}
+		CHKmalloc(field_json = json_object_new_string(value));
```

Comment:
Do we always want a JSON string? I would have expected numbers to be JSON numbers and booleans to be JSON booleans. I didn't investigate the previous code closely. This might be easier to handle if you converted directly to the JSON values in the switch statement rather than using an intermediate value.

Reply:
You're right, thanks for the pointers.